hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/batching/PartitionAwareCallableBatchingRunnable.java

```diff
@@ -0,0 +1,130 @@
+ * PartitionAwareCallable on all partitions this thread is responsible for.
+ *
+ * PartitionAwareCallable will be created using the PartitionAwareCallableFactory class.
+ * THere will be one PartitionAwareCallable per all partitions the runnable will run the processing for.
+ *
+ * IMPORTANT:
+ * <ul>
+ * <li>Runnable assumes to be run on a partition-thread. If not run on a partition-thread a RuntimeException will be thrown.</li>
+ * <li>This optimization highly relies on the threading-model and has been implemented to speed up invocations that span all
+ * partitions, e.g. HD-index-query, or potentially many more like IMap.size(), etc.</li>
+ * </ul>
+ *
+ * @see PartitionAwareCallable
+ * @see PartitionAwareCallableFactory
+ * @since 3.9
+ */
+public class PartitionAwareCallableBatchingRunnable implements Runnable {
+
+    private final PartitionAwareCallableFactory factory;
+    private final IPartitionService partitionService;
+    private final Address memberAddress;
+    private final int partitionThreadCount;
+
+    private final CopyOnWriteArrayList results;
+    private final AtomicInteger finished;
+    private final ResultFuture resultFuture;
+
+    /**
+     * @param nodeEngine nodeEngine
+     * @param factory    factory of PartitionAwareCallable to be run on the partition thread.
+     */
+    public PartitionAwareCallableBatchingRunnable(NodeEngine nodeEngine, PartitionAwareCallableFactory factory) {
+        this.factory = factory;
+        this.partitionService = nodeEngine.getPartitionService();
+        this.memberAddress = nodeEngine.getThisAddress();
+        this.partitionThreadCount = ((OperationServiceImpl) nodeEngine.getOperationService()).getPartitionThreadCount();
+
+        this.finished = new AtomicInteger(0);
+        this.resultFuture = new ResultFuture(nodeEngine, nodeEngine.getLogger(getClass()));
+        this.results = new CopyOnWriteArrayList();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void run() {
+        if (!(Thread.currentThread() instanceof PartitionOperationThread)) {
+            throw new IllegalArgumentException("The PartitionAwareCallableBatchingRunnable has to be run on a partition-thread");
+        }
+        PartitionOperationThread currentThread = ((PartitionOperationThread) Thread.currentThread());
+        int partitionThreadId = currentThread.getThreadId();
+        try {
+            int currentPartitionId = partitionThreadId;
+            IPartition[] partitions = partitionService.getPartitions();
+            runSequentially(currentThread, currentPartitionId, partitions);
+        } finally {
+            int value = finished.incrementAndGet();
+            if (!resultFuture.isDone() && value == partitionThreadCount) {
+                resultFuture.setResult(results);
+            }
+        }
+    }
+
+    private void runSequentially(PartitionOperationThread currentThread, int currentPartitionId, IPartition[] partitions) {
```

Comment:
how will this behave in case of partition migrations? Any wrong results? 
Also this sequential run seems can effectively block other partition-operations from running for a long time if we scan a big chunk of memory(i can see you did also mention interleaving)? Is this ok?

Reply:
Partitions will be missed and the result won't have it in the result. The query engine handles it by spawning separate partition operations for missed partitions. No change here with the new mechanism.